api/src/paths/project/{projectId}/survey/{surveyId}/sample-site/delete.ts

```diff
@@ -0,0 +1,154 @@
+          discriminator: 'ProjectPermission'
+        },
+        {
+          validSystemRoles: [SYSTEM_ROLE.DATA_ADMINISTRATOR],
+          discriminator: 'SystemRole'
+        }
+      ]
+    };
+  }),
+  deleteSurveySampleSiteRecords()
+];
+
+POST.apiDoc = {
+  description: 'Delete survey sample sites.',
+  tags: ['survey'],
+  security: [
+    {
+      Bearer: []
+    }
+  ],
+  parameters: [
+    {
+      in: 'path',
+      name: 'projectId',
+      schema: {
+        type: 'integer',
+        minimum: 1
+      },
+      required: true
+    },
+    {
+      in: 'path',
+      name: 'surveyId',
+      schema: {
+        type: 'integer',
+        minimum: 1
+      },
+      required: true
+    }
+  ],
+  requestBody: {
+    description: 'Survey sample site delete request object.',
+    content: {
+      'application/json': {
+        schema: {
+          type: 'object',
+          required: ['surveySampleSiteIds'],
+          properties: {
+            surveySampleSiteIds: {
+              items: {
+                type: 'integer',
+                minimum: 1
+              }
+            }
+          }
+        }
+      }
+    }
+  },
+  responses: {
+    204: {
+      description: 'Delete survey sample site OK'
+    },
+    400: {
+      $ref: '#/components/responses/400'
+    },
+    401: {
+      $ref: '#/components/responses/401'
+    },
+    403: {
+      $ref: '#/components/responses/403'
+    },
+    500: {
+      $ref: '#/components/responses/500'
+    },
+    default: {
+      $ref: '#/components/responses/default'
+    }
+  }
+};
+
+export function deleteSurveySampleSiteRecords(): RequestHandler {
+  return async (req, res) => {
+    const surveyId = Number(req.params.surveyId);
+    const surveySampleSiteIds = req.body.surveySampleSiteIds as number[];
+
+    if (!surveySampleSiteIds) {
+      throw new HTTP400('Missing required body `surveySampleSiteIds`');
+    }
+
+    const connection = getDBConnection(req['keycloak_token']);
+
+    try {
+      await connection.open();
+
+      const observationService = new ObservationService(connection);
+      const sampleLocationService = new SampleLocationService(connection);
+
+      const sampleLocations = await sampleLocationService.getSampleLocationsForSurveyId(surveyId);
+
+      const unableToDelete: number[] = [];
+
+      const deleteSampleSites = surveySampleSiteIds.map(async (surveySampleSiteId) => {
+        if (
+          (await observationService.getObservationsCountBySampleSiteId(surveyId, surveySampleSiteId)).observationCount >
+          0
+        ) {
+          unableToDelete.push(surveySampleSiteId);
+        } else {
+          await sampleLocationService.deleteSampleLocationRecord(surveySampleSiteId);
+        }
+      });
+
+      await Promise.all(deleteSampleSites);
+
+      const notDeleted = sampleLocations.filter((item) => unableToDelete.includes(item.survey_sample_site_id));
+
+      if (notDeleted.length > 0) {
```

Comment:
If we know that all the sample sites in `unableToDelete` will fail deletion, we should do this check before performing the actual delete. In other words we don't learn anything we didn't already know when performing the delete.